### PR TITLE
Addition of StFwdTrack and StFwdTrackCollection to StEvent

### PR DESCRIPTION
**Pull request by jdbrice**
Builds on FWD upgrade PR 396 (BFC additions)

cons +StRoot/StEvent succeeds but just cons fails with inability to find the StPtrVecFwdTrack based on new StFwdTrack class

### CODE REVIEW

**Comment by genevb** [comment]
Commit [2052ab7](https://github.com/star-bnl/star-sw/pull/397/commits/2052ab789919c60fcba13b2fd2f02ea5674537b8) overlaps with #396 ?

**Comment by jdbrice** [comment]
> Commit [2052ab7](https://github.com/star-bnl/star-sw/pull/397/commits/2052ab789919c60fcba13b2fd2f02ea5674537b8) overlaps with #396 ?

Yes, three PRs in a row that are sub branches of each other. 

396: Big full chain
397: StEvent additions
398: FCS Track match maker

If i understand git properly, after the first (396) is merged to main, the later ones will have those changes disappear but will inherit any changes we make during this review

File: StRoot/StEvent/StFwdTrack.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        return sqrt( cov[8] );
+    }
+};
+
+class StFwdTrack : public StObject {
+
+public:
+    StFwdTrack( genfit::Track * );
+    vector<StFwdTrackProjection> mProjections;
+    genfit::Track *mGenfitTrack;
```
Comment (plexoos) [inline StRoot/StEvent/StFwdTrack.h:57]: Let's wait for Jason to chime in but I think this dependency contradicts the original concept of StEvent

File: StRoot/StEvent/StFwdTrack.cxx [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    return fitStatus->getNFailedPoints();
+}
+
+const double StFwdTrack::chi2() const {
+    if (!mGenfitTrack)
+        return false;
+    // uses the default track rep
+    auto fitStatus = mGenfitTrack->getFitStatus();
+    if ( !fitStatus ) 
+        return false;
```
Comment (veprbl) [inline StRoot/StEvent/StFwdTrack.cxx:59]: Return `false` from double? Should those checks be moved to constructor/factory?

File: StRoot/StEvent/StFwdTrack.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    const unsigned int   numberOfFitPoints() const;
+    // const unsigned int   numberOfPossibleFitPoints() const;
+
+    // Number of points used in the track seed step
+    // const unsigned int   numberOfSeedPoints() const;
+
+    // momentum at the primary vertex
+    const StThreeVectorD momentum() const;
+    const StThreeVectorD momentumAt(int _id = 1) const;
+    const char charge() const;
```
Comment (veprbl) [inline StRoot/StEvent/StFwdTrack.h]: Too much const
```suggestion
    bool   didFitConverge() const;
    bool   didFitConvergeFully() const;
    int    numberOfFailedPoints() const;
    double chi2() const;
    double ndf() const;
    double pval() const;

    // error on pT upper and lower 1-sigma values
    // add access to cov matrix

    // Number of fit points used by GenFit
    unsigned int   numberOfFitPoints() const;
    // unsigned int   numberOfPossibleFitPoints() const;

    // Number of points used in the track seed step
    // unsigned int   numberOfSeedPoints() const;

    // momentum at the primary vertex
    StThreeVectorD momentum() const;
    StThreeVectorD momentumAt(int _id = 1) const;
    char charge() const;
```

File: StRoot/StEvent/StFwdTrack.cxx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        id = id -1;
+        if ( id < 0 ){
+            LOG_WARN << " Cannot get momentum at point: " << _id << endm;
+            return StThreeVectorD( 0, 0, 0 );
+        }
+    } // while valid point
+
+    auto cr = mGenfitTrack->getCardinalRep();
+    TVector3 p = cr->getMom(mGenfitTrack->getFittedState(id, cr));
+    LOG_INFO << "Momentum at point: " << id << endm;
```
Comment (veprbl) [inline StRoot/StEvent/StFwdTrack.cxx:116]: Sounds like `LOG_DEBUG` at most.

File: StRoot/StEvent/StFwdTrack.h (hunk 1 above)
Comment (jdbrice) [reply StRoot/StEvent/StFwdTrack.h:57]: OK, but just a reminder. We discussed the dependency issue in a few S&C meetings, and I implemented this after those discussions

File: StRoot/StEvent/StFwdTrack.h (hunk 3 above)
Comment (jdbrice) [reply StRoot/StEvent/StFwdTrack.h]: why would you remove the const qualifier?

Comment (jdbrice) [reply StRoot/StEvent/StFwdTrack.h]: Do you mean it is redundant when the function is a const?

Comment (veprbl) [reply StRoot/StEvent/StFwdTrack.h]: You pass POD (bool,char,int,double,...), not by reference, so they are always copied by value, so const qualifier doesn't mean anything (there is nothing a caller can modify).

The StThreeVectorD is also not a container (i.e., we assume, it doesn't hold references inside), so, again, you have it copied all of its information over -- there is nothing to modify (unless you want to return `const StThreeVectorD&`, which is fine by me).

Comment (veprbl) [reply StRoot/StEvent/StFwdTrack.h]: So, yes, it's redundant, but for a different reason.

File: StRoot/StEvent/StFwdTrack.h (hunk 1 above)
Comment (klendathu2k) [reply StRoot/StEvent/StFwdTrack.h:57]: Chiming in...  We don't add external package dependencies to StEvent. 

But looking at StFwdTrack.h... other than exposing genfit::Track as public... why do you need to store the genfit track?
It looks like you're just calculating / returning values which can be computed in StEventFiller and saved in member variables of StFwdTrack.

File: StRoot/StEvent/StFwdTrack.h (hunk 3 above)
Comment (plexoos) [reply StRoot/StEvent/StFwdTrack.h]: Thank you @veprbl for the patch. I am committing it and resolving

**Comment by starsdong** [comment]
Hi Daniel, just want to follow up on this? Based on our discussion two weeks ago
https://drupal.star.bnl.gov/STAR/event/2022/11/09/SC-management-meeting
the plan is that you will update this PR following the new strategy. So do we expect this to come soon before we take further action on this PR? Thanks

**Comment by starsdong** [comment]
This PR is superseded by PR#492.

**Pull request closed without merging**